pkg/gameserversets/controller_test.go

```diff
@@ -622,6 +622,98 @@ func TestSyncGameServerSet(t *testing.T) {
 		assert.True(t, updated, "A game servers should have been updated")
 	})
 
+	t.Run("adding and deleting errored gameservers", func(t *testing.T) {
+		gsSet := defaultFixture()
+		list := createGameServers(gsSet, 5)
+
+		// make some as unhealthy
+		list[0].Annotations = map[string]string{agonesv1.GameServerErroredAtAnnotation: time.Now().Add(-10 * time.Second).UTC().Format(time.RFC3339)}
+		list[0].Status.State = agonesv1.GameServerStateError
+
+		updated := false
+		count := 0
+
+		c, m := newFakeController()
+		m.AgonesClient.AddReactor("list", "gameserversets", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			return true, &agonesv1.GameServerSetList{Items: []agonesv1.GameServerSet{*gsSet}}, nil
+		})
+		m.AgonesClient.AddReactor("list", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			return true, &agonesv1.GameServerList{Items: list}, nil
+		})
+
+		m.AgonesClient.AddReactor("update", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			ua := action.(k8stesting.UpdateAction)
+			gs := ua.GetObject().(*agonesv1.GameServer)
+			assert.Equal(t, gs.Status.State, agonesv1.GameServerStateShutdown)
+
+			updated = true
+			assert.Equal(t, "test-0", gs.GetName())
+			return true, nil, nil
+		})
+		m.AgonesClient.AddReactor("create", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			ca := action.(k8stesting.CreateAction)
+			gs := ca.GetObject().(*agonesv1.GameServer)
+
+			assert.True(t, metav1.IsControlledBy(gs, gsSet))
+			count++
+			return true, gs, nil
+		})
+
+		ctx, cancel := agtesting.StartInformers(m, c.gameServerSetSynced, c.gameServerSynced)
+		defer cancel()
+
+		c.syncGameServerSet(ctx, gsSet.ObjectMeta.Namespace+"/"+gsSet.ObjectMeta.Name) // nolint: errcheck
+
+		assert.Equal(t, 6, count)
+		assert.True(t, updated, "A game servers should have been updated")
+	})
+
+	t.Run("adding and delay deleting errored gameservers", func(t *testing.T) {
+		gsSet := defaultFixture()
+		list := createGameServers(gsSet, 5)
+
+		// make some as unhealthy
+		list[0].Annotations = map[string]string{agonesv1.GameServerErroredAtAnnotation: time.Now().UTC().Format(time.RFC3339)}
```

Comment:
Non blocking suggestion:

If it's useful, could create an instance of https://pkg.go.dev/k8s.io/utils/clock (we do this in a few places in the codebase), so you can fake the passage of time in testing.

For example:
https://github.com/googleforgames/agones/blob/85dd69ab8a3f26a41cfed368a4f843e1473cbb91/pkg/fleetautoscalers/controller.go#L80

If not useful, feel free to resolve and ignore.

Reply:
I dont think it is important for this change, as the time stuff is minor and easy to test, but if it got more complex, I definitely see it as a good idea.

Reply:
Sounds reasonable. Resolving.